src/Adapters/Laravel/CollisionServiceProvider.php

```diff
@@ -60,14 +61,9 @@ public function register()
                 return new Provider(null, $handler);
             });
 
-            $appExceptionHandler = $this->app->make(ExceptionHandlerContract::class);
-
-            $this->app->singleton(
-                ExceptionHandlerContract::class,
-                function ($app) use ($appExceptionHandler) {
-                    return new ExceptionHandler($app, $appExceptionHandler);
-                }
-            );
+            $this->app->extend(ExceptionHandlerContract::class, function (ExceptionHandlerContract $exceptionHandler, Container $container) {
```

Comment:
Can you add some tests to explain what exactly this is solving vs the existing behaviour?

Reply:
Not sure how we could test this. 

The problem with the old code arises when the exception handler's constructor has a dependency, like this:

```php
public function __construct(Foo $foo) {}
```

And `Foo` is provided by some other service provider - say, `FooServiceProvider`. Now, when `CollisionServiceProvider::register()` is called before `FooServiceProvider::register()`, `Foo` won't be registered in the container yet and `$this->app->make(ExceptionHandlerContract::class)` will fail.
This can be worked around by disabling package auto-discovery and manually loading the providers in the right order, but that's obviously not optimal.

The new code uses `Container:.extend()` instead, which is the recommended way to extend/wrap things after they are resolved from the container only.

Reply:
A quick way to test it locally in Laravel 6+
Not sure how to get such a case in a test though

```php
$app->singleton(
    Illuminate\Contracts\Debug\ExceptionHandler::class,
    function () use ($app) {
        $app->make('mailer');

        return $app->make(DefaultHandler::class);
    }
);
```

in `bootstrap/app.php`